More changes to environments (#2144)
chef/chef-web-docs: chef_master/source/environments.rst

.. tag environment

An environment is a way to map an organization's real-life workflow to what can be configured and managed when using Chef Infra. Every organization begins with a single environment called the ``_default`` environment, which cannot be modified (or deleted). Additional environments can be created to reflect each organization's patterns and workflow. For example, creating ``production``, ``staging``, ``testing``, and ``development`` environments.

.. end_tag

The _default Environment
=====================================================

> Every Chef Infra Server organization must have at least one environment. Each organization starts out with a single environment that is named ``_default``. The ``_default`` environment cannot be modified in any way. Nodes, roles, run-lists, cookbooks (and cookbook versions), and attributes specific to an organization can only be associated with a custom environment.

Reviewer: ```suggestion
Every Chef Infra Server organization must have at least one environment. Each organization starts out with a single ``_default`` environment. The ``_default`` environment cannot be modified in any way. Nodes, roles, run-lists, cookbooks (and cookbook versions), and attributes specific to an organization can only be associated with a custom environment.
```

> Environment Attributes

Reviewer: This section comes out of the blue. ELIF why we're suddenly discussing attributes.

=====================================================
.. tag environment_attribute

> Attributes can be defined in an environment and then used to override the default attributes in a cookbook. When an environment is applied during a Chef Infra Client run, environment attributes are compared to the attributes that are already present on the node. When the environment attributes take precedence over the default attributes, Chef Infra Client applies those new settings and values during a Chef Infra Client run.

Reviewer: Defining attributes in an environment lets you override the default attributes in a cookbook. When a Chef Infra Client run applies an environment, the client compares the environment attributes to the attributes that are already defined on the node [IN THE COOKBOOK? WHY ARE WE SUDDENLY TALKING ABOUT NODES?] Environment attributes set to ``override`` take precedence over the default attributes set on the node by cookbooks. Chef Infra Client applies the ``override`` environment attributes and values during a Chef Infra Client run.

...or something like this

> Environment attributes can be set to either ``default`` attribute level or an ``override`` attribute level.

Reviewer: Maybe an example would help?

       .. end_tag

> Attribute Precedence

Reviewer: "Precedence" is a much better word! 👍

-----------------------------------------------------

> Environments are one of several locations where attributes can be applied in Chef Infra. It's important to understand how the precedence level works in order to understand, which attributes will be applied when Chef Infra Client runs.

Reviewer: Environments are one of several different places [FILES?] for defining attributes in Chef Infra. Attributes are defined in cookbook attribute files, recipes, environments, roles, and also in Ohai.  Attribute _precedence_ explains why an attribute defined in one file wins out over another during a Chef Infra Client run and gets applied to a node.

Reviewer: I think we should reverse the order of the attribute precedence list, so #1 is the Ohai automatic attribute.

.. end_tag

Cookbook Pinning
=====================================================

> Cookbook versions can be pinned per evironment. This allows for the controlled rollout of new cookbook releases through environments for testing before new versions reach production environments. See the environment format examples below for the cookbook pinning syntax.

Reviewer: ```suggestion
Cookbook versions can be pinned in each environment, which allows you to control the rollout of new cookbook releases through successive testing environments before releasing new cookbook versions into production environments. See the environment format examples below for the cookbook pinning syntax.
```

Environment Formats
=====================================================

> Environments may be stored on disk (any in source control) in two formats:

Reviewer: ```suggestion
Store environments on disk (in any source control) in one of two formats:
```

Author: We said may because it's entirely optional to store them on disk. It's the right thing to do, but still optional.

Reviewer: I'll think about different words.

  .. note:: After the environment has been set via bootstrap, the environment is set in the client.rb file and may not be modified using the Chef management console or the ``edit`` argument of the ``knife node`` subcommand.
* By setting the ``environment`` configuration entry in the client.rb file ; when Chef Infra Client runs, it will pick up the value and then set the ``chef_environment`` attribute of the node

Move Nodes
-----------------------------------------------------

> Nodes can be moved between environments, such as from a "dev" to a "production" environment by using the ``knife exec`` subcommand. For example:

Reviewer: ```suggestion
Use the ``knife exec`` subcommand to move nodes between environments, such as from a "dev" to a "production" environment. For example:
```
